pkgs/development/libraries/openmpi/default.nix

```diff
@@ -131,7 +129,12 @@ stdenv.mkDerivation (finalAttrs: {
     (lib.withFeatureAs fabricSupport "ofi" (lib.getDev libfabric))
     # The flag --without-ofi-libdir is not supported from some reason, so we
     # don't use lib.withFeatureAs
-  ] ++ lib.optionals fabricSupport [ "--with-ofi-libdir=${lib.getLib libfabric}/lib" ];
+  ] ++ lib.optional fabricSupport "--with-ofi-libdir=${lib.getLib libfabric}/lib"
+    # The pmix flags should only be set when pmix is used
+    ++ lib.optionals stdenv.isLinux [
+    "--with-pmix=${lib.getDev pmix}"
```

Comment:
I disagree with this change - you still want to set `--with-pmix=internal` no?

Reply:
Is there a need to set this at all on darwin? "internal" should be the default if it can be built on a given platform. Otherwise it should not be built at all.  Note, that pmix is feature to simplify startup of a large number process over a multiple machines. Unless you have a large cluster of darwin machines pmix is not of much use anyway.

Reply:
> Is there a need to set this at all on darwin? "internal" should be the default if it can be built on a given platform. Otherwise it should not be built at all.

Wait, are you saying that there is a difference between using `--with-pmix=internal` and not using at all?

Reply:
This is the help text from configure:
```
  --with-pmix(=DIR)       Build pmix support. DIR can take one of three
                          values: "internal", "external", or a valid directory
                          name. "internal" forces Open MPI to use its internal
                          copy of pmix. "external" forces Open MPI to use an
                          external installation of pmix. Supplying a valid
                          directory name also forces Open MPI to use an
                          external installation of pmix, and adds DIR/include,
                          DIR/lib, and DIR/lib64 to the search path for
                          headers and libraries. Note that Open MPI no longer
                          supports --without-pmix. If no argument is
                          specified, Open MPI will search default locations
                          for pmix and fall back to an internal version if one
                          is not found.
  --with-pmix-libdir=DIR  Search for pmix libraries in DIR. Should only be
                          used if an external copy of pmix is being used.
```

I would not request to build pmix at all on non-Linux, but let openmpi's internal defaults decide if it wants to build it or not.

Reply:
> I would not request to build pmix at all on non-Linux, but let openmpi's internal defaults decide if it wants to build it or not.

Usually in packages with many such build options that I maintain, I prefer to track upstream's default behavior and mimic it in the Nix functional behavior ("a function always returns something" sort of). What's weird to me, is that there shouldn't be a significant difference between the bundled pmix and the one we package. Perhaps we should just try to enable `pmix` for all platforms?

Reply:
> Perhaps we should just try to enable pmix for all platforms?

What would be the benefit? pmix requires tight integration with a workload manager. That is already difficult enough to achieve on Linux. Why would we add the extra maintenance effort if it is a niche (if any at all) application on non-Linux.

Reply:
> > Perhaps we should just try to enable pmix for all platforms?
> 
> What would be the benefit? pmix requires tight integration with a workload manager. That is already difficult enough to achieve on Linux. Why would we add the extra maintenance effort if it is a niche (if any at all) application on non-Linux.

The benefit would be that you'd handle all platforms consistently. If there will be a bug in the future with `pmix` only on 1 platform, you'd notice that and be able to focus on that when building `pmix`, and not when building a `openmpi`. Also, it'd be easier to apply patches to `pmix` when it is not a submodule. Note how they simply use a git submodule for pmix:

https://github.com/open-mpi/ompi/tree/v5.0.3/3rd-party

Reply:
To be honest, I can not see that benefit. For me this just means more maintenance effort, to make sure this also builds on darwin (where it is not even needed). Do we know if openmpi even builds with the internal pmix on Darwin?

Reply:
> To be honest, I can not see that benefit. For me this just means more maintenance effort, to make sure this also builds on darwin.

You are making sure it also builds fine on Darwin when you are building `openmpi` on Darwin, because you are building it from the submodule.

> ... (where it is not even needed). Do we know if openmpi even builds with the internal pmix on Darwin?

You know it builds with both, because they are the same - up to minor mismatches between the `pmix` release _we_ build, and the one that was used as the submodule when `openmpi` was released.

Reply:
> You know it builds with both, because they are the same

What I actually meant is: does openmpi build the internal pmix if no `--with-pmix` is specified or is it simply disabled on darwin per default?

Reply:
> > You know it builds with both, because they are the same
> 
> What I actually meant is: does openmpi build the internal pmix if no `--with-pmix` is specified or is it simply disabled on darwin per default?

I suspect that if `--with-pmix` is not specified, they simply use the bundled pmix. Perhaps @nim65s could share with us the build log and we will be more sure.